internal/expr/query.go

```diff
@@ -147,19 +174,41 @@ func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error)
 			continue
 		}
 		if idx >= len(qe.outputs) {
-			return nil, fmt.Errorf("internal error: sqlair column not in outputs (%d>=%d)", idx, len(qe.outputs))
-		}
-		field := qe.outputs[idx]
-		outputVal, ok := typeDest[field.structType]
-		if !ok {
-			return nil, fmt.Errorf("type %q found in query but not passed to get", field.structType.Name())
+			return nil, nil, fmt.Errorf("internal error: sqlair column not in outputs (%d>=%d)", idx, len(qe.outputs))
+		}
+		typeMember := qe.outputs[idx]
+		switch tm := typeMember.(type) {
+		case structField:
+			outputVal, ok := typeDest[tm.structType]
+			if !ok {
+				return nil, nil, fmt.Errorf("type %q found in query but not passed to get", tm.structType.Name())
+			}
+			val := outputVal.FieldByIndex(tm.index)
+			if !val.CanSet() {
+				return nil, nil, fmt.Errorf("internal error: cannot set field %s of struct %s", tm.name, tm.structType.Name())
+			}
+			ptrs = append(ptrs, val.Addr().Interface())
+		case mapKey:
+			mapDecodeInfo, ok := mapDecodeInfos[tm.mapType]
+			if !ok {
+				return nil, nil, fmt.Errorf("type %q found in query but not passed to get", tm.mapType.Name())
+			}
+			// Scan in to a new value with the type of the maps value
+			mapDecodeInfo.keys = append(mapDecodeInfo.keys, tm.name)
+			mapDecodeInfo.values = append(mapDecodeInfo.values, reflect.New(tm.mapType.Elem()).Elem())
+			ptrs = append(ptrs, mapDecodeInfo.values[len(mapDecodeInfo.values)-1].Addr().Interface())
 		}
+	}
+	var mapInfoSlice = []*MapDecodeInfo{}
+	for _, v := range mapDecodeInfos {
+		mapInfoSlice = append(mapInfoSlice, v)
+	}
+	return ptrs, mapInfoSlice, nil
```

Comment:
Perhaps as a possible way out of the API issue here, we can define this and return as the second result here. This logic agrees with the other comments above, in that most of the logic is shared between structs and maps.
```
onSuccess := func() {
	if !hasMap {
		return // Save time for the common case.
	}
	for pos, column := range columns {
		idx, ok := markerIndex(column)
		if !ok {
			continue
		}
		typeMember, ok := qe.outputs[idx]
		if !ok {
			continue
		}
		tm, ok := typeMember.(mapKey); ok {
		if !ok {
			continue
		}
		outputVal := typeDest[tm.mapType]
		outputVal.SetMapIndex(reflect.ValueOf(tm.name, ptrs[pos])
	}
}
```

Reply:
I've used the idea of closures but instead of repeating the logic for checking the columns I've created a closure for each `SetMapIndex`, put these in a list and then had `onSuccess` iterate through and call each of these closures. This works pretty well because it really minimises the repeated code.

Let me know if you would still rather it done the way you've given here though.

Reply:
Changed to save the information in a struct rather than a list of closures